File: filefields.c
#define _XOPEN_SOURCE 600

#include <assert.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include "buf.h"
#include "field.h"
#include "file.h"
#include "filefields.h"
#include "group.h"
#include "logging.h"
#include "options.h"
#include "string.h"
#include "user.h"

char *humanbytes(unsigned long bytes);
void printnametobuf(File *file, Options *options, Buf *buf);

/**
 * Dynamically allocate a formatted string (portable asprintf replacement).
 * Returns a malloc'd string or NULL on failure. Caller must free.
 */
static char *xasprintf(const char *fmt, ...)
{
    va_list ap;
    va_start(ap, fmt);
    int n = vsnprintf(NULL, 0, fmt, ap);
    va_end(ap);
    if (n < 0) return NULL;
    char *s = malloc(n + 1);
    if (!s) return NULL;
    va_start(ap, fmt);
    vsnprintf(s, n + 1, fmt, ap);
    va_end(ap);
    return s;
}

/**
 * Dynamically allocate a strftime result.
 * Returns a malloc'd string or NULL on failure. Caller must free.
 */
static char *xstrftime(const char *fmt, const struct tm *tm)
{
    size_t size = 64;
    for (;;) {
        char *buf = malloc(size);
        if (!buf) return NULL;
        size_t n = strftime(buf, size, fmt, tm);
        if (n > 0) return buf;
        free(buf);
        size *= 2;
        if (size > 4096) return NULL;
    }
}

FieldList *getfilefields(File *file, Options *options)
{
    if (file == NULL) {
        errorf("file is NULL\n");
        return NULL;
    }

    List *fieldlist = newlist();
    if (fieldlist == NULL) {
        errorf("fieldlist is NULL\n");
        return NULL;
    }

    /*
     * with -L, display information about the link target file by setting file = target
     * the originally named file is saved as "link" for displaying the name
     */
    File *link = file;
    if (options->targetinfo && islink(file)) {
        /* XXX ls seems to use stat() instead of lstat() here */
        file = getfinaltarget(file);
        if (!file) {
            errorf("Cannot determine target of %s\n", getname(link));
            /* file is NULL,
               code below should handle this and print "?" or similar */
        }
    }

    if (options->size) {
        Field *field = getsizefield(file, options);
        if (!field) goto error;
        append(field, fieldlist);
    }

    if (options->inode) {
        Field *field = getinodefield(file, options);
        if (!field) goto error;
        append(field, fieldlist);
    }

    if (options->modes) {
        Field *field = getmodesfield(file, options);
        if (!field) goto error;
        append(field, fieldlist);
    }

    if (options->linkcount) {
        Field *field = getlinkfield(file, options);
        if (!field) goto error;
        append(field, fieldlist);
    }

    if (options->owner) {
        Field *field = getownerfield(file, options);
        if (!field) goto error;
        append(field, fieldlist);
    }

    if (options->group) {
        Field *field = getgroupfield(file, options);
        if (!field) goto error;
        append(field, fieldlist);
    }

    if (options->perms) {
        Field *field = getpermsfield(file, options);
        if (!field) goto error;
        append(field, fieldlist);
    }

    if (options->bytes) {
        Field *field = getbytesfield(file, options);
        if (!field) goto error;
        append(field, fieldlist);
    }

    if (options->datetime) {
        Field *field = getdatetimefield(file, options);
        if (!field) goto error;
        append(field, fieldlist);
    }

    Field *field = getnamefield(link, options);
    append(field, fieldlist);

    return (FieldList *)fieldlist;

error:
    freelist(fieldlist, (free_func)freefield);
    return NULL;
}

Field *getbytesfield(File *file, Options *options)
{
    char *s;
    if (!isstat(file)) {
        s = xasprintf("?");
    } else {
        long bytes = getsize(file);
        if (options->sizestyle == SIZE_HUMAN) {
            s = humanbytes(bytes);
        } else {
            if (options->sizestyle != SIZE_DEFAULT) {
                errorf("Unknown sizestyle %d\n", options->sizestyle);
            }
            s = xasprintf("%ld", bytes);
        }
    }
    if (!s) return NULL;
    Field *field = newfield(s, ALIGN_RIGHT, strlen(s));
    free(s);
    return field;
}

Field *getdatetimefield(File *file, Options *options)
{
    char *s = NULL;
    if (isstat(file)) {
        time_t timestamp;
        switch (options->timetype) {
        case TIME_ATIME:
            timestamp = getatime(file);
            break;
        case TIME_CTIME:
            timestamp = getctime(file);
            break;
        case TIME_BTIME:
            timestamp = getbtime(file);
            break;
        default:
            errorf("Unknown time attribute\n");
            /* fall through */
        case TIME_MTIME:
            timestamp = getmtime(file);
            break;
        }
        struct tm *timestruct = localtime(&timestamp);
        if (!timestruct) {
            errorf("timestruct is NULL\n");
            return NULL;
        }
        if (options->timeformat != NULL) {
            s = xstrftime(options->timeformat, timestruct);
        } else if (options->timestyle == TIME_RELATIVE) {
            assert(options->now > 0);
            time_t seconds_ago = options->now - timestamp;
            if (seconds_ago > 60*60*24*31*12) {
                s = xasprintf("%ld years", seconds_ago/60/60/24/31/12);
            } else if (seconds_ago > 60*60*24*31) {
                s = xasprintf("%ld months", seconds_ago/60/60/24/31);
            } else if (seconds_ago > 60*60*24) {
                s = xasprintf("%ld days", seconds_ago/60/60/24);
            } else if (seconds_ago > 60*60) {
                s = xasprintf("%ld hours", seconds_ago/60/60);
            } else if (seconds_ago > 60) {
                s = xasprintf("%ld minutes", seconds_ago/60);
            } else if (seconds_ago >= 0) {
                s = xasprintf("%ld seconds", seconds_ago);
            } else {
                s = xstrftime("%b %e  %Y", timestruct);
            }
        } else {
            /* month day hour and minute if file was modified in the last 6 months,
               month day year otherwise */
            assert(options->now > 0);
            s = xstrftime("%b %e  %Y", timestruct);
            if (timestamp <= options->now) {
                if (timestamp > options->now - 6*86400) {
                    free(s);
                    s = xstrftime("%a    %H:%M", timestruct);
                } else if (timestamp > options->now - 180*86400) {
                    free(s);
                    s = xstrftime("%b %e %H:%M", timestruct);
                }
            }
        }
    } else {
        s = xasprintf("?");
    }
    if (!s) return NULL;
    Field *field = newfield(s, ALIGN_RIGHT, strlen(s));
    free(s);
    return field;
}

Field *getgroupfield(File *file, Options *options)
{
    char *s;
    if (isstat(file)) {
        gid_t gid = getgroupnum(file);
        if (options->numeric) {
            s = xasprintf("%lu", (unsigned long)gid);
        } else {
            char *groupname = get(options->groupnames, gid);
            if (!groupname) {
                groupname = getgroupname(gid);
                if (!groupname) {
                    groupname = xasprintf("%lu", (unsigned long)gid);
                    if (!groupname) return NULL;
                    set(options->groupnames, gid, groupname);
                    free(groupname);
                } else {
                    set(options->groupnames, gid, groupname);
                }
                groupname = get(options->groupnames, gid);
            }
            s = xasprintf("%s", groupname);
        }
    } else {
        s = xasprintf("?");
    }
    if (!s) return NULL;
    Field *field = newfield(s, ALIGN_LEFT, strlen(s));
    free(s);
    return field;
}

Field *getinodefield(File *file, Options *options)
{
    char *s;
    if (isstat(file)) {
        ino_t inode = getinode(file);
        s = xasprintf("%lu", inode);
    } else {
        s = xasprintf("?");
    }
    if (!s) return NULL;
    Field *field = newfield(s, ALIGN_RIGHT, strlen(s));
    free(s);
    return field;
}

Field *getlinkfield(File *file, Options *options)
{
    char *s;
    if (isstat(file)) {
        nlink_t nlinks = getlinkcount(file);
        s = xasprintf("%lu", (unsigned long)nlinks);
    } else {
        s = xasprintf("?");
    }
    if (!s) return NULL;
    Field *field = newfield(s, ALIGN_RIGHT, strlen(s));
    free(s);
    return field;
}

Field *getmodesfield(File *file, Options *options)
{
    char *s;
    if (isstat(file)) {
        s = getmodes(file);
    } else {
        s = xasprintf("???????????");
    }
    if (!s) return NULL;
    Field *field = newfield(s, ALIGN_LEFT, strlen(s));
    free(s);
    return field;
}

/**
 * Print the file name with flags and colors if requested
 * flags, color start, filename, color end, and flags are made into a single field
 * since we don't want spaces in between
 * if -L is given and the file is a symlink, the returned fields include the name
 * of the link's target (multiple times if there are multiple links)
 * i.e. link -> file
 * or even link -> link -> link -> file
 *
 * A symlink loop will be printed as
 * link1 -> link2 -> link3 -> link1
 */
Field *getnamefield(File *file, Options *options)
{
    if (!file) {
        errorf("file is NULL\n");
        return NULL;
    }
    if (!options) {
        errorf("options is NULL\n");
        return NULL;
    }

    Buf *buf = newbuf();
    if (!buf) {
        errorf("buf is NULL\n");
        return NULL;
    }

    /* print the file itself... */
    printnametobuf(file, options, buf);

    if (options->showlinks) {
        Map *linkmap = newmap();
        if (linkmap) {
            /* resolve and print symlink targets recursively */
            while (isstat(file) && islink(file)) {
                if (inmap(linkmap, getinode(file))) {
                    /* error already printed by getfilefields */
                    /* no file to stat, but want to print the name field */
                    file = NULL;
                    break;
                } else {
                    set(linkmap, (uintmax_t)getinode(file), NULL);
                }
                file = gettarget(file);
                bufappend(buf, " -> ", 4, 4);
                printnametobuf(file, options, buf);
            }
            freemap(linkmap);
        } else {
            errorf("Out of memory?\n");
        }
    } else if (options->showlink) {
        /* print only the first link target without stat'ing the target */
        if (isstat(file) && islink(file)) {
            file = gettarget(file);
            bufappend(buf, " -> ", 4, 4);
            printnametobuf(file, options, buf);
        }
    }

    enum align align = ALIGN_NONE;
    if (options->displaymode != DISPLAY_ONE_PER_LINE) {
        align = ALIGN_LEFT;
    }

    Field *field = newfield(bufstring(buf), align, bufscreenpos(buf));
    if (field == NULL) {
        errorf("field is NULL\n");
        return NULL;
    }
    freebuf(buf);
    return field;
}

Field *getownerfield(File *file, Options *options)
{
    char *s;
    if (isstat(file)) {
        uid_t uid = getownernum(file);
        if (options->numeric) {
            s = xasprintf("%lu", (unsigned long)uid);
        } else {
            char *username = get(options->usernames, uid);
            if (!username) {
                username = getusername(uid);
                if (!username) {
                    username = xasprintf("%lu", (unsigned long)uid);
                    if (!username) return NULL;
                    set(options->usernames, uid, username);
                    free(username);
                } else {
                    set(options->usernames, uid, username);
                }
                username = get(options->usernames, uid);
            }
            s = xasprintf("%s", username);
        }
    } else {
        s = xasprintf("?");
    }
    if (!s) return NULL;
    Field *field = newfield(s, ALIGN_LEFT, strlen(s));
    free(s);
    return field;
}

Field *getpermsfield(File *file, Options *options)
{
    char *perms = getperms(file);
    if (perms == NULL) {
        errorf("perms is NULL\n");
        return NULL;
    }
    int width = strlen(perms);
    Field *field = newfield(perms, ALIGN_RIGHT, width);
    free(perms);
    return field;
}

Field *getsizefield(File *file, Options *options)
{
    char *s;
    if (!isstat(file)) {
        s = xasprintf("?");
    } else {
        unsigned long blocks = getblocks(file, options->blocksize);
        if (options->sizestyle == SIZE_HUMAN) {
            unsigned long bytes = blocks * options->blocksize;
            s = humanbytes(bytes);
        } else {
            if (options->sizestyle != SIZE_DEFAULT) {
                errorf("Unknown sizestyle %d", options->sizestyle);
            }
            s = xasprintf("%lu", blocks);
        }
    }
    if (!s) return NULL;
    Field *field = newfield(s, ALIGN_RIGHT, strlen(s));
    free(s);
    return field;
}

void printnametobuf(File *file, Options *options, Buf *buf)
{
    assert(file != NULL);
    assert(options != NULL);
    assert(buf != NULL);

    /*
     * print a character *before* the file showing its type (-O)
     *
     * early versions of BSD printed directories like [this]
     */
    switch (options->flags) {
    case FLAGS_NORMAL:
        break;
    case FLAGS_OLD:
        if (!isstat(file))
            bufappend(buf, "?", 1, 1);
        else if (isdir(file))
            bufappend(buf, "[", 1, 1);
        else if (islink(file))
            bufappend(buf, "@", 1, 1);
        else if (isexec(file))
            bufappend(buf, "*", 1, 1);
        break;
    case FLAGS_NONE:
        break;
    }

    /* color the file (-G and -K) */
    /* these escape sequences shouldn't move the cursor,
     * hence the 4th arg to bufappend is 0 */
    /* TODO change this to something like setcolor(COLOR_BLUE) */
    int colorused = 0;
    if (options->color) {
        if (!isstat(file)) {
            bufappend(buf, options->colors->red, strlen(options->colors->red), 0);
            colorused = 1;
        } else if (isdir(file)) {
            bufappend(buf, options->colors->blue, strlen(options->colors->blue), 0);
            colorused = 1;
        } else if (islink(file)) {
            bufappend(buf, options->colors->cyan, strlen(options->colors->cyan), 0);
            colorused = 1;
        } else if (isexec(file)) {
            bufappend(buf, options->colors->green, strlen(options->colors->green), 0);
            colorused = 1;
        }
    }

    const char *name = getname(file);
    printtobuf(name, options->escape, buf);
    /* don't free name */

    /* reset the color back to normal (-G and -K) */
    if (colorused) {
        bufappend(buf, options->colors->none, strlen(options->colors->none), 0);
    }

    /* print a character after the file showing its type (-F and -O) */
    switch (options->flags) {
    case FLAGS_NORMAL:
        if (!isstat(file))
            bufappend(buf, "?", 1, 1);
        else if (isdir(file))
            bufappend(buf, "/", 1, 1);
        else if (islink(file))
            bufappend(buf, "@", 1, 1);
        else if (isfifo(file))
            bufappend(buf, "|", 1, 1);
        else if (issock(file))
            bufappend(buf, "=", 1, 1);
        else if (isexec(file))
            bufappend(buf, "*", 1, 1);
        else
            bufappend(buf, " ", 1, 1);
        break;
    case FLAGS_OLD:
        if (!isstat(file))
            bufappend(buf, "?", 1, 1);
        else if (isdir(file))
            bufappend(buf, "]", 1, 1);
        else if (islink(file))
            bufappend(buf, "@", 1, 1);
        else if (isexec(file))
            bufappend(buf, "*", 1, 1);
        else
            bufappend(buf, " ", 1, 1);
        break;
    case FLAGS_NONE:
        break;
    }
}

char *humanbytes(unsigned long bytes)
{
    char *units[] = {"B", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB"};
    int nunits = sizeof(units) / sizeof(units[0]);
    double dbytes = bytes;
    int i;
    for (i = 0; dbytes >= 1000 && i < nunits - 1; i++) {
        dbytes /= 1000;
    }
    /* round to the nearest integer using %.0f */
    return xasprintf("%.0f %s", dbytes, units[i]);
}

File: rust-l/src/filefields.rs
use crate::color::Colors;
use crate::field::{Align, Field};
use crate::file::File;
use crate::options::*;

pub fn get_file_fields(file: &File, options: &Options) -> Vec<Field> {
    let mut fields = Vec::new();

    // If target_info and file is a symlink, we show info about the target
    // but display the original name
    let target_stat = if options.target_info && file.is_link() {
        file.get_final_target_stat()
    } else {
        None
    };

    if options.size {
        fields.push(get_size_field(file, options));
    }
    if options.inode {
        fields.push(get_inode_field(file));
    }
    if options.modes {
        fields.push(get_modes_field(file));
    }
    if options.link_count {
        fields.push(get_link_count_field(file));
    }
    if options.owner {
        fields.push(get_owner_field(file, options));
    }
    if options.group {
        fields.push(get_group_field(file, options));
    }
    if options.perms {
        fields.push(get_perms_field(file));
    }
    if options.bytes {
        fields.push(get_bytes_field(file, options, &target_stat));
    }
    if options.datetime {
        fields.push(get_datetime_field(file, options, &target_stat));
    }

    fields.push(get_name_field(file, options));

    fields
}

fn get_size_field(file: &File, options: &Options) -> Field {
    let blocks = file.get_blocks(options.blocksize);
    Field::right(format!("{}", blocks))
}

fn get_inode_field(file: &File) -> Field {
    Field::right(format!("{}", file.get_inode()))
}

fn get_modes_field(file: &File) -> Field {
    Field::left(file.get_modes())
}

fn get_link_count_field(file: &File) -> Field {
    Field::right(format!("{}", file.get_nlink()))
}

fn get_owner_field(file: &File, options: &Options) -> Field {
    let uid = file.get_uid();
    let name = if options.numeric {
        format!("{}", uid)
    } else {
        get_username(uid).unwrap_or_else(|| format!("{}", uid))
    };
    Field::left(name)
}

fn get_group_field(file: &File, options: &Options) -> Field {
    let gid = file.get_gid();
    let name = if options.numeric {
        format!("{}", gid)
    } else {
        get_groupname(gid).unwrap_or_else(|| format!("{}", gid))
    };
    Field::left(name)
}

fn get_perms_field(file: &File) -> Field {
    Field::right(file.get_perms())
}

fn get_bytes_field(file: &File, options: &Options, target_stat: &Option<libc::stat>) -> Field {
    let size = if let Some(st) = target_stat {
        st.st_size
    } else {
        file.get_size()
    };

    let s = match options.size_style {
        SizeStyle::Human => human_bytes(size as u64),
        SizeStyle::Default => format!("{}", size),
    };
    Field::right(s)
}

fn get_datetime_field(file: &File, options: &Options, target_stat: &Option<libc::stat>) -> Field {
    let time = if let Some(st) = target_stat {
        match options.time_type {
            TimeType::Mtime => st.st_mtime,
            TimeType::Atime => st.st_atime,
            TimeType::Ctime => st.st_ctime,
        }
    } else {
        match options.time_type {
            TimeType::Mtime => file.get_mtime(),
            TimeType::Atime => file.get_atime(),
            TimeType::Ctime => file.get_ctime(),
        }
    };

    let s = match options.time_style {
        TimeStyle::Iso => format_time_iso(time),
        TimeStyle::Relative => format_time_relative(time, options.now),
        TimeStyle::Traditional => format_time_traditional(time, options.now),
    };
    Field::right(s)
}

fn get_name_field(file: &File, options: &Options) -> Field {
    let mut buf = String::new();
    let colors = if options.color { Colors::new() } else { None };

    // Opening flag for old-style
    if options.flags == FlagStyle::Old {
        if file.is_dir() {
            buf.push('[');
        } else if file.is_exec() {
            buf.push('*');
        }
    }

    // Color start
    if let Some(ref c) = colors {
        if file.is_dir() {
            buf.push_str(&c.blue);
        } else if file.is_link() {
            buf.push_str(&c.cyan);
        } else if file.is_exec() {
            buf.push_str(&c.green);
        }
    }

    // The name itself, with escaping applied
    let name = escape_name(file.name(), options.escape);
    buf.push_str(&name);

    // Color end
    if let Some(ref c) = colors {
        if file.is_dir() || file.is_link() || file.is_exec() {
            buf.push_str(&c.none);
        }
    }

    // Closing flag for old-style
    if options.flags == FlagStyle::Old {
        if file.is_dir() {
            buf.push(']');
        } else if file.is_exec() {
            buf.push('*');
        }
    }

    // Normal flags (-F)
    if options.flags == FlagStyle::Normal {
        if file.is_dir() {
            buf.push('/');
        } else if file.is_link() {
            buf.push('@');
        } else if file.is_exec() {
            buf.push('*');
        } else if file.is_fifo() {
            buf.push('|');
        } else if file.is_sock() {
            buf.push('=');
        }
    }

    // Symlink display
    if options.show_links && file.is_link() {
        let chain = file.get_link_chain();
        for target_name in &chain {
            buf.push_str(" -> ");
            buf.push_str(&escape_name(target_name, options.escape));
        }
    } else if options.show_link && file.is_link() {
        if let Some(target) = file.get_target() {
            buf.push_str(" -> ");
            buf.push_str(&escape_name(target.name(), options.escape));
        }
    }

    // Calculate visible width (excluding escape sequences)
    let width = visible_width(&buf);
    Field::new(buf, Align::Left, width)
}

fn escape_name(name: &str, escape: EscapeMode) -> String {
    match escape {
        EscapeMode::None => name.to_string(),
        EscapeMode::Question => {
            name.chars()
                .map(|c| {
                    if c.is_control() {
                        '?'
                    } else {
                        c
                    }
                })
                .collect()
        }
        EscapeMode::C => {
            let mut s = String::new();
            for c in name.chars() {
                match c {
                    '\\' => s.push_str("\\\\"),
                    '\n' => s.push_str("\\n"),
                    '\t' => s.push_str("\\t"),
                    '\r' => s.push_str("\\r"),
                    '\x07' => s.push_str("\\a"),
                    '\x08' => s.push_str("\\b"),
                    '\x0c' => s.push_str("\\f"),
                    '\x0b' => s.push_str("\\v"),
                    '\x1b' => s.push_str("\\033"),
                    c if c.is_control() => {
                        s.push_str(&format!("\\{:03o}", c as u32));
                    }
                    c => s.push(c),
                }
            }
            s
        }
    }
}

fn visible_width(s: &str) -> usize {
    let mut width = 0;
    let mut in_escape = false;
    for c in s.chars() {
        if in_escape {
            if c.is_ascii_alphabetic() {
                in_escape = false;
            }
        } else if c == '\x1b' {
            in_escape = true;
        } else {
            width += unicode_width::UnicodeWidthChar::width(c).unwrap_or(0);
        }
    }
    width
}

fn human_bytes(bytes: u64) -> String {
    const UNITS: &[&str] = &["B", "KB", "MB", "GB", "TB", "PB", "EB"];
    let mut size = bytes as f64;
    for unit in UNITS {
        if size < 1024.0 {
            return format!("{:.0}{}", size, unit);
        }
        size /= 1024.0;
    }
    format!("{:.0}ZB", size)
}

fn format_time_traditional(time: i64, now: i64) -> String {
    use std::fmt::Write;
    let tm = unsafe {
        let t = time as libc::time_t;
        let mut tm: libc::tm = std::mem::zeroed();
        libc::localtime_r(&t, &mut tm);
        tm
    };

    let months = [
        "Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec",
    ];
    let month = months[tm.tm_mon as usize % 12];
    let day = tm.tm_mday;

    // If within ~6 months, show time; otherwise show year
    let six_months = 180 * 24 * 3600;
    let mut s = String::new();
    if (now - time).abs() < six_months {
        let _ = write!(s, "{} {:>2} {:02}:{:02}", month, day, tm.tm_hour, tm.tm_min);
    } else {
        let _ = write!(s, "{} {:>2}  {}", month, day, tm.tm_year + 1900);
    }
    s
}

fn format_time_iso(time: i64) -> String {
    let tm = unsafe {
        let t = time as libc::time_t;
        let mut tm: libc::tm = std::mem::zeroed();
        libc::localtime_r(&t, &mut tm);
        tm
    };
    format!(
        "{:04}-{:02}-{:02} {:02}:{:02}:{:02}",
        tm.tm_year + 1900,
        tm.tm_mon + 1,
        tm.tm_mday,
        tm.tm_hour,
        tm.tm_min,
        tm.tm_sec
    )
}

fn format_time_relative(time: i64, now: i64) -> String {
    let diff = now - time;
    if diff < 0 {
        return "future".to_string();
    }

    let seconds = diff;
    let minutes = seconds / 60;
    let hours = minutes / 60;
    let days = hours / 24;
    let months = days / 30;
    let years = days / 365;

    if years > 0 {
        format!("{} year{}", years, if years == 1 { "" } else { "s" })
    } else if months > 0 {
        format!("{} month{}", months, if months == 1 { "" } else { "s" })
    } else if days > 0 {
        format!("{} day{}", days, if days == 1 { "" } else { "s" })
    } else if hours > 0 {
        format!("{} hour{}", hours, if hours == 1 { "" } else { "s" })
    } else if minutes > 0 {
        format!("{} minute{}", minutes, if minutes == 1 { "" } else { "s" })
    } else {
        format!(
            "{} second{}",
            seconds,
            if seconds == 1 { "" } else { "s" }
        )
    }
}

fn get_username(uid: u32) -> Option<String> {
    unsafe {
        let pw = libc::getpwuid(uid);
        if pw.is_null() {
            None
        } else {
            let name = std::ffi::CStr::from_ptr((*pw).pw_name);
            Some(name.to_string_lossy().to_string())
        }
    }
}

fn get_groupname(gid: u32) -> Option<String> {
    unsafe {
        let gr = libc::getgrgid(gid);
        if gr.is_null() {
            None
        } else {
            let name = std::ffi::CStr::from_ptr((*gr).gr_name);
            Some(name.to_string_lossy().to_string())
        }
    }
}
